README.md
# useMediaQuery and useElementQuery
[![Version](https://img.shields.io/npm/v/use-element-query.svg)](https://www.npmjs.com/package/use-element-query)
[![NPM-Size](https://flat.badgen.net/bundlephobia/minzip/use-element-query)](https://www.npmjs.com/package/use-element-query)
[![Build Status](https://travis-ci.org/Merri/use-element-query.svg)](https://travis-ci.org/Merri/use-element-query)

These React hooks make it easy for you to match with your media queries.

**Get it**: `npm install use-element-query --save`

## What?

Given things you want to do with your media queries...

```jsx
const queries = {
    '(max-width: 299px)': { thing: 'A' },
    '(min-width: 300px) and (max-width: 599px)': { thing: 'B' },
    '(min-width: 600px)': { thing: 'C' }
}
```

... you can ask for a list of things that match!

```jsx
function ResponsiveComponent() {
    const things = useMediaQuery(queries)
    // server side: = []
    // client side: = [{ thing: 'A' }]
    //          or: = [{ thing: 'B' }]
    //          or: = [{ thing: 'C' }]
    return (
        <div>
            Things that match:
            <pre>{JSON.stringify(things, null, 4)}</pre>
        </div>
    )
}
```

As the output is an array you can have multiple matching queries at one time. Also, you can use whatever you want as
your "thing": it can be a string, object, array, React component, CSS styles, callback function...

`useMediaQuery` works with `window.matchMedia` so you can use any query that works with it.

## What about `useElementQuery`?

This is an extension that creates `<object data="about:blank" type="text/html" />` element inside container element that
is determinted via `ref`. A `position` that is not `static` is required as the `object` element is made as large as the
container using `position: absolute`. This is a [trick from 2013 used for element resize detection](http://www.backalleycoder.com/2013/03/18/cross-browser-event-based-element-resize-detection/).
This library however instead uses `matchMedia` of the `object` which then allows to use media queries for a viewport
that is the same size as the container element!

### Usage

```jsx
import React, { useMemo } from 'react'
import { useElementQuery } from 'use-element-query'

// define this outside component to keep the same reference
// or if you need this to be dynamic: useMemo
const queries = {
    '(max-width: 299px)': 'small',
    '(min-width: 300px) and (max-width: 599px)': 'medium',
    '(min-width: 600px)': 'large'
}

function ResponsiveComponent() {
    // third value is always an array with all matching items
    const [ref, baseStyle, [size = 'default']] = useElementQuery(queries)

    // for demo purposes set a width that is smaller than the viewport width
    const style = useMemo(() => ({ width: '50%', ...baseStyle }), [baseStyle])

    // ref and baseStyle always passed to the container
    return (
        <div ref={ref} style={style}>
            <h2>Size is <code>{size}</code></h2>
        </div>
    )
}
```

### Can `useElementQuery` be used with Styled Components?

Yes, [check out this CodeSandbox](https://codesandbox.io/s/demo-useelementquery-with-styled-components-ccy44)!

You will end up with a bit more boilerplate code than with `:container()` syntax that is used by solutions like
[Styled Container Query](https://www.npmjs.com/package/styled-container-query), but this library is much smaller yet
also supports changing your render path entirely depending on available container width; although one could argue you
should always keep the HTML/DOM the same and only change the appearance via CSS.

### Caveats

While this trick is performant there are some memory consumption issues so you should avoid using this on very many
elements at the same time. Also, `object` element is added to the DOM and this may have some effect to your CSS if you
need `:first-child` or `:last-child` on the container. The current minimalist design tries to make usage as easy as
possible and thus `object` is added without React: if this proves to be problematic it might be necessary the change the
API at some point.

Circularity issues are shared with all solutions that do container queries these days, and is the reason there are
container query libraries instead of a CSS standard.

## Motivation

I got annoyed at work how hard it was to calculate "content width" in CSS, especially with a dynamic sidebar and having
complex product lists. So originally this was a quick weekend project to get something usable that gives CSS /
`matchMedia` style of syntax via a React hook. I didn't find any other solution that would keep familiar media query
syntax, or that would allow for anything to be used as a result value from query. Also, I do not know of any other
solution that has used `matchMedia` from a `<object data="about:blank" type="text/html" />`.

Container queries and element queries have been long in the discussion, but ultimately it seems they might never be
actually implemented despite numerous proposals, polyfills, and other JavaScript based implementations.

### Alternatives for Container Queries

Minzipped sizes are as of 2019-11-28.

- [React Container Query](https://www.npmjs.com/package/react-container-query) (6,5 kB)
- [React Element Query](https://www.npmjs.com/package/react-element-query) (8,7 kB)
- [Styled Container Query](https://www.npmjs.com/package/styled-container-query) (5,6 kB)
- [ZeeCoder: React Container Query](https://www.npmjs.com/package/@zeecoder/react-container-query) (16,1 kB)

## Note about the status of the repo

I really should upgrade to latest Babel and stuff, instead of copy-pasting from old projects. Packages are a bit old.
Also, should see the trouble to learn Storybook and make use of it.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2019-11-28

- Add `useMediaQuery`
- Reuse `matchMedia` objects


## [1.0.0] - 2019-11-25

- Add `useElementQuery`


src/useQueries.js
import { useCallback, useEffect, useState } from 'react'

export default function useQueries(mediaQueries) {
    const [activeItems, setActiveItems] = useState([])

    const updateQueries = useCallback(() => {
        const nextActiveItems = mediaQueries.reduce((nextActiveItems, item) => {
            if (item.mq.matches) nextActiveItems.push(item.value)
            return nextActiveItems
        }, [])
        if (
            nextActiveItems.length !== activeItems.length ||
            nextActiveItems.some((activeItem, index) => activeItems[index] !== activeItem)
        ) {
            setActiveItems(nextActiveItems)
        }
    }, [activeItems, mediaQueries])

    useEffect(() => {
        if (mediaQueries.length === 0) return
        updateQueries()
        mediaQueries.forEach(item => {
            item.mq.addListener(updateQueries)
        })
        return () => {
            mediaQueries.forEach(item => {
                item.mq.removeListener(updateQueries)
            })
        }
    }, [mediaQueries, updateQueries])

    return activeItems
}


src/useElementQuery.js
// yet another variant of http://www.backalleycoder.com/2013/03/18/cross-browser-event-based-element-resize-detection/
import { useCallback, useMemo, useState } from 'react'

import useQueries from './useQueries'

const isIE = 'navigator' in global && navigator.userAgent.match(/Trident/)

const registry = new WeakMap()
let objectStyle = 'document' in global && document.getElementById('object-data-about-blank')

function init() {
    if (objectStyle) return
    objectStyle = document.createElement('style')
    objectStyle.id = 'object-data-about-blank'
    objectStyle.textContent = `
object[data="about:blank"] {
    clip: rect(0 0 0 0);
    height: 100%;
    left: 0;
    pointer-events: none;
    position: absolute;
    top: 0;
    user-select: none;
    width: 100%;
}
`
    document.head.appendChild(objectStyle)
}

function createObject(element, setElement) {
    const object = document.createElement('object')
    object.setAttribute('aria-hidden', 'true')
    object.onload = function() {
        const object = this
        function complete() {
            if (!object.contentDocument) setTimeout(complete, 50)
            else setElement(object.contentDocument.defaultView)
        }
        complete()
    }
    object.tabIndex = -1
    object.type = 'text/html'
    if (!isIE) object.data = 'about:blank'
    element.appendChild(object)
    if (isIE) object.data = 'about:blank'
}

const staticToRelativeOverride = { position: 'relative' }

function initContainer(element, setElement, setStyle) {
    const style = getComputedStyle(element)
    let destroyed = false
    let styleApplied = false
    function applyStyle() {
        if (styleApplied || style.position === '') return
        if (style.position === 'static') setStyle(staticToRelativeOverride)
        styleApplied = true
    }
    applyStyle()
    createObject(element, viewElement => {
        if (destroyed) return
        applyStyle()
        if (viewElement) setElement(viewElement)
    })
    return () => {
        setElement(null)
        setStyle(null)
        destroyed = true
    }
}

export default function useElementQuery(queries = {}) {
    const [storage] = useState(() => new Map())
    const [ref, setRef] = useState(null)
    const [element, setElement] = useState(null)
    const [style, setStyle] = useState(null)

    const getRef = useCallback(
        nextRef => {
            init()
            if (ref === nextRef) return
            if (registry.has(ref)) registry.get(ref)()
            if (nextRef) {
                registry.set(nextRef, initContainer(nextRef, setElement, setStyle))
            }
            setRef(nextRef)
        },
        [ref]
    )

    const elementQueries = useMemo(() => {
        if (!element || !('matchMedia' in element)) return []
        return Object.keys(queries).map(key => {
            if (!storage.has(key)) storage.set(key, element.matchMedia(key))
            return { mq: storage.get(key), value: queries[key] }
        })
    }, [element, queries])

    return [getRef, style, useQueries(elementQueries)]
}
